icu4j/main/core/src/test/java/com/ibm/icu/dev/test/lang/UnicodeSetTest.java

```diff
@@ -3198,4 +3205,178 @@ public void TestPatternCodePointComplement() {
+                fromUnicodeSetIterator.add(stringToString(usi.getString()));
+            }
+        }
+        assertEquals("strings :: UnicodeSetIterator", expected.toString(), fromUnicodeSetIterator.toString());
+
+        StringJoiner fromStream = newResult();
+        us.stringStream().map(UnicodeSetTest::stringToString).forEach(fromStream::add);
+        assertEquals("strings :: stringStream", expected.toString(), fromStream.toString());
+    }
+
+    @Test
+    // Iterates on strings AND on code points (converted to strings)
+    public void testIterationMethodsCodepointsAndStrings() {
+        final UnicodeSet us = populateUnicodeSet();
+
+        // Build the reference test result
+        StringJoiner expect = newResult();
+        for (String str : us) {
+            expect.add(stringToString(str));
+        }
+
+        StringJoiner fromStream = newResult();
+        us.stream().map(UnicodeSetTest::stringToString).forEach(fromStream::add);
+
+        assertEquals("code points + strings :: stream", expect.toString(), fromStream.toString());
+    }
+
+    /*
+     * Helper methods for testing various iterations.
+     * The main goal is to make the results of any possible failures more readable
+     * by formatting code points to something like U+03A3(Σ) and wrapping strings in double quotes.
+     */
+
+    private static String codePointToString(int codepoint) {
+        String fromCodePoint = new String(Character.toChars(codepoint));
+        return String.format("U+%04X(%s)", codepoint, fromCodePoint);
+    }
+
+    private static String rangeToString(EntryRange range) {
+        return rangeString(range.codepoint, range.codepointEnd);
+    }
+
+    private static String rangeString(int codepoint, int codepointEnd) {
+        return String.format("%s-%s", codePointToString(codepoint), codePointToString(codepointEnd));
+    }
+
+    private static String stringToString(String str) {
+        return String.format("\"%s\"", str);
+    }
+
+    private static StringJoiner newResult() {
+        return new StringJoiner(", ");
+    }
+
+    @Test
+    public void testParallelStreams() {
+        if (!isVerbose()) {
+            return;
+        }
+        UnicodeSet us = UnicodeSet.ALL_CODE_POINTS;
+
+        long start = System.nanoTime();
+        int sumNormal = us.codePointStream().sum();
+        long timeNormal = System.nanoTime() - start;
+        System.out.println("codePointStream.normal   : " + timeNormal);
+
+        start = System.nanoTime();
+        int sumParallel = us.codePointStream().parallel().sum();
+        long timeParallel = System.nanoTime() - start;
+        System.out.println("codePointStream.parallel : " + timeParallel);
+
+        // On my machines this is about 1.2, so it is a bit faster.
+        // Unlikely to have any practical benefit, this is more to test that the
+        // parallel stream works the same as the normal one, by comparing the sums.
```

Comment:
Why is there any difference if trySplit() always returns null? Doesn't that make it sequential?

Reply:
After implementing an optimized version of codepoint iteration it is even faster (close to 1.5)